### PR TITLE
Array reactivity fix

### PR DESCRIPTION
**Pull request by joshhanley**
This fix ensures that if a store prop value contains an array is replaced with another array, that it still has reactivity and any watchers that were setup still remain.

I have added a test which demonstrates the issue.

Hope this helps!

### CODE REVIEW

**Comment by ryangjchandler** [comment]
Wicked, thanks @joshhanley. Killing it with the OSS contributions!

**Comment by joshhanley** [comment]
@ryangjchandler no worries 🙂  Thanks! If you get a chance check your discord DMs, I sent you a link 👀

File: src/observable.js
Comment (ryangjchandler) [inline src/observable.js:26]: I think this entire block should probably be done before updating it with the new value. Setting `.__watchers` at the moment will technically cause a re-render on top of the re-render from `target[key] = value`.

Just need to move this `if` statement / block to before the `target[key] = value` line and replace some of the names, e.g. `target[key].__watchers` probably needs to be `value['__watchers']`.

**Review by ryangjchandler — CHANGES_REQUESTED** [review]

File: src/observable.js
Comment (joshhanley) [reply src/observable.js:26]: @ryangjchandler hmm ok I wasn't aware of the watcher assignment causing re-rendering. Are you able to explain that?

If I remember rightly, I think I ran into an issue with that, where if the watchers were already assigned, when value gets assigned, then the watchers were running twice.

That's why I pulled the `target[key] = value` out of the callbacks set call and put the watcher reassignment in after that, because the watchers then get called in the `callback`.

Let me know what your thoughts are on that, and if you want me to try and put that above anyway, I will do it and run some experiments.

Comment (ryangjchandler) [reply src/observable.js:26]: Perhaps a simpler change here would be disabling reactivity before you re-assign the watchers and then enabling it afterwards. This is what I'm doing as part of the `start` method: https://github.com/ryangjchandler/spruce/blob/1fedc3b29472955e9809ce2ef3e3647b66a6a95d/src/index.js#L59

If you think about it logically, setting `target[key].__watchers` will be assigning a property on a `Proxy` instance and will then be picked up the `set` trap.

Comment (joshhanley) [reply src/observable.js:26]: @ryangjchandler I've added a test which simulates the two way watchers (which currently passes) so we can experiment, refactor and ensure it all still passes. I can remove it later once it's all working before you're ready to merge if you don't want it there.

I've experimented with disabling the reactivity around the watcher reassignment and that didn't work, I was able to stop the reactivity but then some of the other watcher tests failed. See code and screenshot below
```js
set(target, key, value, receiver) {
    if (!isNullOrUndefined(value) && (isObject(value) || isArray(value))) {
        value = createObservable(value, callbacks)
    }

    // Copy watchers from the original value if they exist
    if (!isNullOrUndefined(target[key]) && !isNullOrUndefined(target[key].__watchers)) {
        // Change this from the window object to passed reference if it works
        window.Spruce.disableReactivity = true
        value.__watchers = target[key].__watchers
        window.Spruce.disableReactivity = false
    }

    callbacks.set(target, key, (target[key] = value), receiver)

    return true
},
```
image


I also tried moving the watcher assignment up to before the `createObservable` call on line 16 but then that also didn't work. See code below and same error as in screenshot above.

```js
set(target, key, value, receiver) {
    if (!isNullOrUndefined(value) && (isObject(value) || isArray(value))) {
        // Copy watchers from the original value if they exist
        if (!isNullOrUndefined(target[key]) && !isNullOrUndefined(target[key].__watchers)) {
            value.__watchers = target[key].__watchers
        }

        value = createObservable(value, callbacks)
    }

    callbacks.set(target, key, (target[key] = value), receiver)

    return true
},
```

Any other thought's on what I can try?

Comment (joshhanley) [reply src/observable.js:26]: @ryangjchandler have you had a chance to have a look at my comment above? 

Let me know if theres anything I can do 🙂

Comment (ryangjchandler) [reply src/observable.js:26]: Sorry @joshhanley, I'll try and get to this in the week!

Comment (joshhanley) [reply src/observable.js:26]: @ryangjchandler no worries! Thanks again!

Comment (ryangjchandler) [reply src/observable.js:26]: @joshhanley I'm going to go ahead and say that if you've got a test to pass that proves your changes fix the watchers problem, then I'm happy to merge this in.

If people report problems then I can take a look and ping you too.

Comment (joshhanley) [reply src/observable.js:26]: @ryangjchandler ok great 🙂  thanks for your support on this!

Yep happy to help work out any issues if they arise.

**Comment by joshhanley** [comment]
@ryangjchandler I've merged in your latest changes from `2.x` and re-built the assets, so should all be good to go.

**Review by ryangjchandler — APPROVED** [review]

**Pull request merged**
